compiler/rustc_data_structures/src/stack.rs

```diff
@@ -13,5 +13,32 @@ const STACK_PER_RECURSION: usize = 1 * 1024 * 1024; // 1MB
 ///
 /// Should not be sprinkled around carelessly, as it causes a little bit of overhead.
 pub fn ensure_sufficient_stack<R>(f: impl FnOnce() -> R) -> R {
-    stacker::maybe_grow(RED_ZONE, STACK_PER_RECURSION, f)
+    let mut mut_closure = Some(f);
+    let mut retval: Option<R> = None;
+    let retval_ref = &mut retval;
+
+    // This wrapper around `f` achieves two things:
+    // * It converts the `impl FnOnce` to a `dyn FnMut`.
+    //   `dyn` because we want it to not be generic, and
+    //   `FnMut` because we can't pass a `dyn FnOnce` around without boxing it.
+    // * It eliminates the generic return value, by writing it to the stack of this function.
+    //   Otherwise the closure would have to return an unsized value, which isn't possible.
+    let dyn_closure: &mut dyn FnMut() = &mut || {
+        // The closure needs to be owned to be invoked, and with this trick
+        // we "move" it into this borrowed closure.
+        let closure = mut_closure.take().unwrap();
+        *retval_ref = Some(closure());
```

Comment:
Can you avoid this `Option` trick by using a `dyn FnOnce` instead? That might require boxing it though ...

Reply:
Sadly no.
I can't pass a `dyn FnOnce`, because it's unsized and would require boxing,
and I can't pass a `&mut FnOnce` because then I get a "Can't move out of borrowed value" error.
This is the best workaround I've found.